chat-android/src/main/java/com/ably/chat/Messages.kt

```diff
@@ -170,25 +180,96 @@ data class SendMessageParams(
 )
 
 interface MessagesSubscription : Cancellation {
-    suspend fun getPreviousMessages(queryOptions: QueryOptions): PaginatedResult<Message>
+    suspend fun getPreviousMessages(start: Long? = null, end: Long? = null, limit: Int = 100): PaginatedResult<Message>
 }
 
-class DefaultMessages(
+internal class DefaultMessagesSubscription(
+    private val chatApi: ChatApi,
     private val roomId: String,
-    private val realtimeClient: RealtimeClient,
+    private val cancellation: Cancellation,
+    internal val fromSerialProvider: () -> DeferredValue<String>,
+) : MessagesSubscription {
+    override fun cancel() {
+        cancellation.cancel()
+    }
+
+    override suspend fun getPreviousMessages(start: Long?, end: Long?, limit: Int): PaginatedResult<Message> {
+        val fromSerial = fromSerialProvider().await()
+        val queryOptions = QueryOptions(start = start, end = end, limit = limit, orderBy = NewestFirst)
+        return chatApi.getMessages(
+            roomId = roomId,
+            options = queryOptions,
+            fromSerial = fromSerial,
+        )
+    }
+}
+
+internal class DefaultMessages(
+    private val roomId: String,
+    realtimeChannels: AblyRealtime.Channels,
     private val chatApi: ChatApi,
 ) : Messages {
 
+    private var listeners: Map<Messages.Listener, DeferredValue<String>> = emptyMap()
+
+    private var channelStateListener: ChannelStateListener
+
+    private var lock = Any()
+
     /**
      * the channel name for the chat messages channel.
      */
     private val messagesChannelName = "$roomId::\$chat::\$chatMessages"
 
-    override val channel: Channel
-        get() = realtimeClient.channels.get(messagesChannelName, ChatChannelOptions())
+    override val channel: Channel = realtimeChannels.get(messagesChannelName, ChatChannelOptions())
+
+    init {
+        channelStateListener = ChannelStateListener {
+            if (!it.resumed) updateChannelSerialsAfterDiscontinuity()
+        }
+        channel.on(ChannelState.attached, channelStateListener)
```

Comment:
We also need to handle `ChannelEvent.update` here - which is RTL2g in the core feature spec

Reply:
Thanks, nice catch!

Reply:
fixed